Experimental support for statically linked NOMAD

jan-provaznik opened the pull request:
Early experimental support for statically linked NOMAD libraries, NOMAD executable and PyNomad interface.

**Disclaimer**
- This is _early_ and experimental version. It has only been tested on Linux. 
- Hoping the automated integration tests for other platforms pass. The tests only check that the changes have not broken anything.

**Overview**
- I've made severe changes to cmake scripts. The libraries are now built both in static and shared versions. The compilation is done only once to save resources. 
- A side effect is that the targets have been renamed, although additional backwards-compatible renaming (or aliasing) could be done.
- I've updated `CMakeFiles` for examples  to match the new cmake targets.
- I've updated `CMakeFiles` for PyNomad to match the new cmake targets as well.
- I've added a new executable target: statically linked NOMAD executable. Mostly for testing purposes.

---

**Statically linked PyNomad**
I have added a brand new setup script (setup_PyNomadStatic.py) for statically linked PyNomad interface. Its current version only supports Linux. I'll test (and extend) the script on Windows. Currently one can execute the following to build the binary distribution wheel

```
export NOMAD_INSTALL=/path/to/build/release
export NOMAD_VERSION=4.3.1
python3 setup_PyNomadStatic.py bdist_wheel
```
where the exported `NOMAD_INSTALL` variable should point to the place NOMAD was installed to (through `cmake --install`) and `NOMAD_VERSION` is arbitrary version string which should eventually match the release of NOMAD. 

Please note that the `build` directory must be removed before running the setup setup script, especially if the interface was built through `cmake -DBUILD_INTERFACE_PYTHON=ON ...`. Both the static and the dynamic version of the package share identical name (i.e., PyNomad) and this confuses `setuptools` if one or the other version already exists.

jan-provaznik committed:
Experimental support for statically linked NOMAD

Early experimental support for statically linked NOMAD libraries,
NOMAD executable and PyNomad interface.

Made changes to cmake targets,
updated examples and standard PyNomad to match.

jan-provaznik committed:
Fixed CMakeFiles for C interface.

ctribes reviewed (commented):
With a single nomad static library (nomad.a or libnomad.a) we can keep the name nomadUtils, nomadEval and nomadAlgo for dynamic library.
A flag to have CMake (-DStaticBuild=ON) produce the static library or not could be convenient.

Usually, we first do PR to the Develop branch. Once testing and merging in Develop works fine we do a PR Develop->Master.

jan-provaznik committed:
Fixed CMakeFilesLibs and setup_PyNomadStatic.

jan-provaznik committed:
Separated STATIC.

jan-provaznik committed:
Fixed up PyNomad interface

jan-provaznik committed:
Fixed problems with MSVC. Static build now optional.

Triggered with -DWITH_STATIC_BUILD=ON.

jan-provaznik committed:
Merged static libraries into nomadStatic.

Overall,
- added PyNomadStatic cmake taget,
- improved PyNomadStatic setup and added proper support for MSVC.

ctribes committed:
Make PyNomad build work without installing first.

Works on OSX. Need to verify on linux and windows.

jan-provaznik committed:
Static build by default.

Build both static and shared libraries.

Link C interface statically.
Link Python interface statically.
Link nomad executable statically.

jan-provaznik committed:
Fixed MSVC missing exports in CInterface header.

ctribes commented on ext/sgtelib/CMakeListsLibs.txt:
```diff
@@ -0,0 +1,52 @@
+# Build the object first. 
+#
+# It will be built only once, then linked twice to make the shared and the
```
This is only for dynamic library.

jan-provaznik replied:
Incorrect comment remained from previous iteration. Nice catch. Fixing.

ctribes commented on ext/sgtelib/CMakeListsLibs.txt:
```diff
@@ -0,0 +1,52 @@
+  )
+endif()
+
+# Install both variants of the library.
```
Only dynamic library

jan-provaznik replied:
Incorrect comment remained from previous iteration. Nice catch. Fixing.

ctribes commented on ext/sgtelib/CMakeListsLibsStatic.txt:
```diff
@@ -0,0 +1,50 @@
+# Build the object first. 
+#
+# It will be built only once, then linked twice to make the shared and the
```
Static only

jan-provaznik replied:
Incorrect comment remained from previous iteration. Nice catch. Fixing.

ctribes commented on interfaces/CInterface/NomadStdCInterface.h:
```diff
@@ -1,6 +1,15 @@
 #ifndef _NOMADSTDCINTERFACE_H_
 #define _NOMADSTDCINTERFACE_H_
 
+#ifdef _MSC_VER
```
Better like that

jan-provaznik committed:
Cleaned up CMakeLists*, fixed leftover comments.

ctribes committed:
Merge branch 'feature-buildPyNomad-without-install' into feature-experimental-static

ctribes committed:
Make PyNomad build work without installing first.

Works on OSX. Need to verify on linux and windows.
Get Nomad version from nomad_version.hpp.
Only PyNomad static build.

jan-provaznik committed:
Link against CInterface a shared nomad libraries.

ctribes committed:
Merge branch 'feature-experimental-static' of https://github.com/jan-provaznik/bbopt-nomad into feature-experimental-static

jan-provaznik committed:
Updated PyNomad to work with MSVC.

jan-provaznik committed:
Introduced PyPi workflow.

jan-provaznik committed:
Introduced 'next' continuous integration workflow.

ctribes commented on interfaces/PyNomad/tests/test_single_constraint.py:
```diff
@@ -0,0 +1,45 @@
+def blackbox (eval_point):
+    eval_f0 = objective(eval_point)
+    eval_g0 = constraint_one(eval_point)
+    eval_g1 = constraint_one(eval_point)
```
Probably 'constraint_two' here

ctribes replied:
Maybe x_expect (line 40) is changed if constraint_two is used.

jan-provaznik replied:
Aye, good catch. Thanks.

jan-provaznik committed:
Fixed incorrect PyNomad test.

jan-provaznik committed:
Test configuration moved to PyPi workflow.

jan-provaznik committed:
Renamed PyNomadStatic target to PyNomad.
